py/segtree.py

```diff
@@ -0,0 +1,216 @@
+
+    def prod(self, l: int, r: int) -> S:
+        """
+        [l, r)の区間の総積を返す
+        :param int l: 半開区間の開始
+        :param int r: 半壊区間の終端
+        :return S: 総積
+        """
+        left_prod = self.e()
+        right_prod = self.e()
+        l += self.sz
+        r += self.sz
+        while l < r:
+            if l & 1 == 1:
+                left_prod = self.op(left_prod, self.data[l])
+                l += 1
+            if r & 1 == 1:
+                r -= 1
+                right_prod = self.op(self.data[r], right_prod)
+            l >>= 1
+            r >>= 1
+        return self.op(left_prod, right_prod)
+
+    def all_prod(self) -> S:
+        """
+        すべての要素の総積を返す
+        :return S: 総積
+        """
+        return self.data[1]
+
+    def max_right(self, l: int, f: Callable[[S], bool]) -> int:
+        """
+        (r = l or f(prod([l, r))) = True) and (r = n or f(prod([l, r+1))) = False)となるrを返す。fが単調なら、f(prod([l, r))) = Trueとなる最大のrとなる。
+        :param l: 半開区間の開始
+        :param f: 単調な判定関数
+        :return int: 最大のr
+        """
+        assert f(self.e())
+        if l == self.n:
+            return self.n
+        l += self.sz
+        while l % 2 == 0:
+            l >>= 1
+        sm = self.e()
+        while f(self.op(sm, self.data[l])):
+            if l.bit_length() != (l+1).bit_length():
+                return self.n
+            sm = self.op(sm, self.data[l])
+            l += 1
+            while l % 2 == 0:
+                l >>= 1
+        while l < self.sz:
+            if not f(self.op(self.data[l*2])):
+                l *= 2
+            else:
+                sm = self.op(sm, self.data[l*2])
+                l = l*2+1
+        return l-self.sz
+
+    def min_left(self, r: int, f: Callable[[S], bool]) -> int:
+        """
+        (l = 0 or f(prod([l, r))) = True) and (l = r or f(prod([l-1, r))) = False)となるlを返す。fが単調なら、f(prod([l, r))) = Trueとなる最小のlとなる。
+        :param r: 半開区間の終端
+        :param f: 単調な判定関数
+        :return int: 最小のl
+        """
+        assert f(self.e())
+        if r == self.n:
+            return self.n
+        r += self.sz-1
+        while r % 2 == 1:
+            r >>= 1
+        sm = self.e()
+        while f(self.op(sm, self.data[r])):
+            if r.bit_length() != (r-1).bit_length():
+                return 0
+            sm = self.op(sm, self.data[r])
+            r -= 1
+            while r % 2 == 1:
+                r >>= 1
+        while r < self.sz:
+            if not f(self.op(self.data[r*2+1], sm)):
+                r = r*2+1
+            else:
+                sm = self.op(self.data[r*2+1], sm)
+                r *= 2
+        return r+1-self.sz
+
+    def __str__(self) -> str:
+        re: List[str] = []
+        for i in range(self.n):
+            re.append(str(self.data[i+self.sz]))
+        return ' '.join(re)
+
+
+class RMaxQ(SegTree, Generic[S]):
+    def __init__(self, n_or_v: Union[int, List[S]], init: S = float('inf')) -> None:
```

Comment:
```suggestion
    def __init__(self, n_or_v: Union[int, List[S]], init: S = -float('inf')) -> None:
```
逆っぽいです
`float('inf')`遅いけど自分で設定可能にしてくださったのはとても良さそうです

docstringのほうも修正お願いします